cylc/flow/platforms.py

```diff
@@ -366,35 +371,46 @@ def generic_items_match(
     return True
 
 
-def get_host_from_platform(platform, method='random'):
+def get_host_from_platform(
+    platform: Dict[str, Any], badhosts: Optional[Set[str]] = None
+) -> str:
     """Placeholder for a more sophisticated function which returns a host
     given a platform dictionary.
 
     Args:
-        platform (dict):
-            A dict representing a platform.
-        method (str):
-            Name a function to use when selecting hosts from list provided
-            by platform.
-            - 'random' (default): Pick a random host from list
-            - 'first': Return the first host in the list
+        platform: A dict representing a platform.
+        badhosts: A set of hosts Cylc knows to be unreachable.
 
     Returns:
-        hostname (str):
-
-    TODO:
-        Other host selection methods:
-            - Random Selection with check for host availability
-
+        hostname: The name of a host.
     """
-    if method == 'random':
-        return random.choice(platform['hosts'])
-    elif method == 'first':  # noqa: SIM106
-        return platform['hosts'][0]
-    else:
-        raise NotImplementedError(
-            f'method {method} is not a valid input for get_host_from_platform'
+    # Get list of goodhosts:
+    if badhosts:
+        # TODO: Change log level to debug?
+        LOG.warning(
+            'Cylc knows that these hosts are unreachable: \n'
+            f'{", ".join(sorted(badhosts))}'
         )
+        goodhosts = list(set(platform['hosts']) - badhosts)
+    else:
+        goodhosts = platform['hosts']
+
+    # Get the selection method
+    method = platform['selection']['method']
+    methods = {
+        'first': lambda goodhosts: goodhosts[0],
+        'random': random.choice
+    }
+    if not goodhosts:
+        raise NoHostsError(f'Unable to find valid host for {platform["name"]}')
+    else:
+        if method not in methods:
+            raise CylcError(
+                f'method {method} is not a valid input for '
+                'get_host_from_platform'
+            )
+        else:
+            return methods[method](goodhosts)
```

Comment:
Perhaps replace the `LOG.warning` above (the one with the TODO comment) with a more generic debug message something like:

```suggestion
            LOG.debug(f'host selection platform={platform["name"]} method={method} badhosts={badhosts}')
            return methods[method](goodhosts)
```